src/main/cpp/src/json_parser.cuh

```diff
@@ -567,6 +444,118 @@ class json_parser {
     }
   }
 
+  __device__ inline std::pair<int, int> write_string(char const* str_pos,
+                                                     char* copy_destination,
+                                                     escape_style w_style)
+  {
+    if (eof(str_pos)) return std::make_pair(0, 0);
+    char const quote_char = *str_pos;
+    // Records string/field name token utf8 bytes size after unescaped
+    // e.g.: For JSON 4 chars string "\\n", after unescaped, get 1 char '\n'
+    // used by checking the max string length
+    int unescped_string_utf8_bytes = 0;
+    // Records string/field name token utf8 bytes size after escaped
+    // e.g.: 4 chars string "\\n", will write out 4 chars: " \ n "
+    int escped_string_utf8_bytes = 0;
+
+    // write the first " if write style is escaped
+    if (escape_style::ESCAPED == w_style) {
+      escped_string_utf8_bytes++;
+      if (nullptr != copy_destination) { *copy_destination++ = '"'; }
+    }
+
+    // skip left quote char
+    if (!try_skip(str_pos, quote_char)) {
+      return std::make_pair(unescped_string_utf8_bytes, escped_string_utf8_bytes);
+    }
+
+    // scan string content
+    while (!eof(str_pos)) {
+      char c = *str_pos;
+      int v  = static_cast<int>(c);
+      if (c == quote_char) {
+        // path 1: match closing quote char
+        str_pos++;
+
+        // check max str len
+        if (!(max_string_utf8_bytes <= 0 ||
+              (max_string_utf8_bytes > 0 && unescped_string_utf8_bytes <= max_string_utf8_bytes))) {
+          return std::make_pair(unescped_string_utf8_bytes, escped_string_utf8_bytes);
+        }
+
+        // write the end " if write style is escaped
+        if (escape_style::ESCAPED == w_style) {
+          escped_string_utf8_bytes++;
+          if (nullptr != copy_destination) { *copy_destination++ = '"'; }
+        }
+
+        return std::make_pair(unescped_string_utf8_bytes, escped_string_utf8_bytes);
+      } else if (v >= 0 && v < 32) {
+        // path 2: unescaped control char
+
+        // copy if enabled, unescape mode, write 1 char
+        if (copy_destination != nullptr && escape_style::UNESCAPED == w_style) {
+          *copy_destination++ = *str_pos;
+        }
+
+        // copy if enabled, escape mode, write more chars
+        if (escape_style::ESCAPED == w_style) {
+          int escape_chars = escape_char(*str_pos, copy_destination);
+          if (copy_destination != nullptr) copy_destination += escape_chars;
+          escped_string_utf8_bytes += (escape_chars - 1);
+        }
+
```

Comment:
```
        // check match if enabled
        const char* match_str_pos = nullptr;
        if (!try_match_char(match_str_pos, nullptr, *str_pos)) {
          return std::make_pair(unescped_string_utf8_bytes, escped_string_utf8_bytes);
        }
```
Please check this block. `try_match_char(nullptr, nullptr, *str_pos)` ?

Reply:
This is to solve `cannot bind non-const lvalue reference of type 'const char*&' to a value of type 'std::nullptr_t'`